python/lsst/daf/butler/_butler_factory.py

```diff
@@ -0,0 +1,165 @@
+    ``DirectButler`` itself is not currently threadsafe, so this guarantee does
+    not buy you much. See DM-42317.
+    """
+
+    def __init__(self, repositories: Mapping[str, str] | None = None) -> None:
+        if repositories is None:
+            self._repositories = None
+        else:
+            self._repositories = dict(repositories)
+
+        self._factories = ThreadSafeCache[str, _FactoryFunction]()
+        self._initialization_locks = NamedLocks()
+
+    def create_butler(self, *, label: str, access_token: str | None) -> Butler:
+        """Create a Butler instance.
+
+        Parameters
+        ----------
+        label : `str`
+            Label of the repository to instantiate, from the ``repositories``
+            parameter to the `ButlerFactory` constructor or the global
+            repository index file.
+        access_token : `str` | `None`
+            Gafaelfawr access token used to authenticate to a Butler server.
+            This is required for any repositories configured to use
+            `RemoteButler`.  If you only use `DirectButler`, this may be
+            `None`.
+
+        Notes
+        -----
+        For a service making requests on behalf of end users, the access token
+        should normally be a "delegated" token so that access permissions are
+        based on the end user instead of the service. See
+        https://gafaelfawr.lsst.io/user-guide/gafaelfawringress.html#requesting-delegated-tokens
+        """
+        factory = self._get_or_create_butler_factory_function(label)
+        return factory(access_token)
+
+    def _get_or_create_butler_factory_function(self, label: str) -> _FactoryFunction:
+        # We maintain a separate lock per label.  We only want to instantiate
+        # one factory function per label, because creating the factory sets up
+        # shared state that should only exist once per repository.  However, we
+        # don't want other repositories' instance creation to block on one
+        # repository that is slow to initialize.
+        with self._initialization_locks.lock(label):
+            if (factory := self._factories.get(label)) is not None:
+                return factory
+
+            factory = self._create_butler_factory_function(label)
+            return self._factories.set_or_get(label, factory)
+
+    def _create_butler_factory_function(self, label: str) -> _FactoryFunction:
+        config_uri = self._get_config_uri(label)
+        config = ButlerConfig(config_uri)
+        butler_type = config.get_butler_type()
+
+        if butler_type == ButlerType.DIRECT:
+            return _create_direct_butler_factory(config)
+        elif butler_type == ButlerType.REMOTE:
+            return _create_remote_butler_factory(config)
+
+    def _get_config_uri(self, label: str) -> ResourcePathExpression:
+        if self._repositories is None:
+            return ButlerRepoIndex.get_repo_uri(label)
+        else:
+            return self._repositories[label]
```

Comment:
It just throws a KeyError for label not found.  It's important for security that this NOT work for URLs, because these labels are sometimes untrusted user input.

Reply:
KeyError from python is a famously opaque error message.